### PR TITLE
fix address checking used in CLI

### LINKED ISSUE #2008: etherbase and unlock flags result in "Fatal: Invalid account address or index" error
I'm getting the following error when I use the `--unlock` and/or `--etherbase` flags:
```
Fatal: Invalid account address or index: '05ae1d0ca6206c6168b42efcd1fbe0ed144e821b'
```
(I've tried it with and without a `0x` prefix, and with and without single and double quotes.)  If I specify `--etherbase`, it throws that error right away.  If I specify `--unlock` only, I get the following output:
```
$ geth --networkid 10101 --datadir /home/jack/.ethereum-10101 --rpc --unlock "05ae1d0ca6206c6168b42efcd1fbe0ed144e821b" console
I1127 02:08:57.920165    7931 database.go:71] Alloted 16MB cache to /home/jack/.ethereum-10101/chaindata
I1127 02:08:57.927169    7931 database.go:71] Alloted 16MB cache to /home/jack/.ethereum-10101/dapp
I1127 02:08:57.931762    7931 backend.go:159] Protocol Versions: [63 62 61], Network Id: 10101
I1127 02:08:57.932212    7931 backend.go:188] Blockchain DB Version: 3
I1127 02:08:57.934240    7931 blockchain.go:210] Last header: #35477 [74985473…] TD=103745547912
I1127 02:08:57.934302    7931 blockchain.go:211] Last block: #35477 [74985473…] TD=103745547912
I1127 02:08:57.934336    7931 blockchain.go:212] Fast block: #35477 [74985473…] TD=103745547912
I1127 02:08:57.935957    7931 server.go:311] Starting Server
I1127 02:09:00.034221    7931 udp.go:207] Listening, enode://9eb7db1d77509652d74908073f4cb653d113a7865ce675a4d270c284399ea9ed84edad0d8e6583eb810303acabe19cfb1f049f001f780b07d0f2e5aae707b17a@[::]:30303
I1127 02:09:00.034600    7931 server.go:552] Listening on [::]:30303
Fatal: Invalid account address or index: '05ae1d0ca6206c6168b42efcd1fbe0ed144e821b'
```
This output is from the following version (on Ubuntu 14.04):
```
$ geth version
Geth
Version: 1.4.0-unstable
Git Commit: 7dde2b902cf81e90b484b1a48f6d45e0abd10e0f
Protocol Versions: [63 62 61]
Network Id: 1
Go Version: go1.5.1
OS: linux
GOPATH=/home/jack/go
GOROOT=/usr/lib/go
```
This problem started for me at commit 1e806c4c775bd98b224eb0249007502d348e737b.  The commit prior to that (8a44451edfa36ea40da564a2fa7ea905d45440a4) does not compile correctly for me, and before that (9e1d9bff3b18cfa09ba96cd027fa653daae10816) it works fine.

### PR DESCRIPTION
**Pull request by zelig**
fixes  #2008

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Fri Nov 27 11:01:01 UTC 2015_

File: common/types.go [hunk 1]
```diff
@@ -95,10 +95,7 @@ func HexToAddress(s string) Address    { return BytesToAddress(FromHex(s)) }
 // IsHexAddress verifies whether a string can represent a valid hex-encoded
 // Ethereum address or not.
 func IsHexAddress(s string) bool {
-	if len(s) == 2+2*AddressLength && IsHex(s[2:]) {
-		return true
-	}
-	if len(s) == 2*AddressLength && IsHex(s) {
+	if len(s) == 2*AddressLength || len(s) == 2+2*AddressLength && s[:2] == "0x" {
```
Comment (karalabe) [inline common/types.go]: The last two checks should be parenthesized.

**Comment by karalabe** [comment]
I'm not entirely sure this is s a valid fix? The old code checked that with or without a prefix, there are AddressLength number of hex characters. The new code just ignores the characters, which may be now arbitrary. So :-1:

**Comment by karalabe** [comment]
IMHO the correct fix would be to add "0x" to the second if clause (and drop the reslice from the first). Don't drop the IsHex as that ensures the characters are actually valid. Also, just looking at IsHex I see that it actually doesn't check the character validity. Could we add that too? So that I don't get to feed geth 0xabcdefghijkl and have it pass?

File: common/bytes.go [hunk 2]
```diff
@@ -139,8 +144,14 @@ func HasHexPrefix(str string) bool {
...
 
 func IsHex(str string) bool {
-	l := len(str)
-	return l >= 4 && l%2 == 0 && str[0:2] == "0x"
+	return hexRe.MatchString(str)
+}
+
+func NormaliseHex(str string) (string, bool) {
+	if HasHexPrefix(str) {
+		str = str[:2]
```
Comment (karalabe) [inline common/bytes.go]: Didn't you mean `str[2:]` here? :)

**Comment by zelig** [comment]
@karalabe supercededd by #2011

**Pull request closed without merging**